search/includes/classes/concurrency-limiter/class-semaphore-backend.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Automattic\VIP\Search\ConcurrencyLimiter;
+
+require_once __DIR__ . '/backendinterface.php';
+
+/**
+ * Semaphore_Backend is more reliable than APCu_Backend (if the process gets killed, all the acquired locks are automatically released).
+ * However, once the semaphore is created, it is impossible to change its `max_acquire` value (in other words, it will not be possible
+ * to change the number of allowed connections).
+ */
+class Semaphore_Backend implements BackendInterface {
+	/** @var int */
+	private static $instances = 0;
+
+	/** @var resource|false|\SysvSemaphore */
+	private $sem;
+
+	/** @var int */
+	private $increments = 0;
+
+	public function __destruct() {
+		if ( $this->sem && $this->increments > 0 ) {
+			while ( $this->increments > 0 ) {
+				sem_release( $this->sem );
+				--$this->increments;
+			}
+		}
+
+		--self::$instances;
+		if ( defined( 'WP_TESTS_DOMAIN' ) && $this->sem && 0 === self::$instances ) {
+			sem_remove( $this->sem );
+		}
```

Comment:
Semaphores are not deleted automatically; they persist even if the process that created them, died. For tests, we remove what we create so that we don't create too many semaphores (in case the tests are run on the live system, not in a container).